Rubberduck.VBEEditor/SafeComWrappers/VBA/VBComponents.cs

```diff
@@ -243,10 +253,21 @@ private static void OnComponentRenamed(VB.VBComponent vbComponent, string oldNam
             if (handler != null)
             {
                 var component = new VBComponent(vbComponent);
-                var project = component.Collection.Parent;
+                IVBProject project;
+                using (var components = component.Collection)
+                {
+                    project = components.Parent;
+                }
                 if (project.Protection != ProjectProtection.Locked)
                 {
-                    handler.Invoke(component, new ComponentRenamedEventArgs(project.ProjectId, project, new VBComponent(vbComponent), oldName));
+                    handler.Invoke(component,
+                        new ComponentRenamedEventArgs(project.ProjectId, project, new VBComponent(vbComponent),
+                            oldName));
+                }
+                else
+                {
+                    project.Dispose();
+                    component.Dispose();
```

Comment:
Does those `Dispose()` really belong in the `else`? I think they should be unconditional? Maybe wrap in a `try/finally` to guarantee cleanup?

Reply:
They belong there because the handlers of the events should receive an indisposed component in the event arguments. Otherwise the event arguments may be unusable. However, in case we do not handle the event at all, we can dispose immediately and this is the only place that has a reference, except the COM safe now.